### PR TITLE
SO-4786: Add support for relationship concrete values

### PR DESCRIPTION
**Pull request by apeteri**

### CODE REVIEW

File: snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedRelationshipRestSearch.java [hunk 1]
```diff
@@ -44,6 +46,15 @@
...
 	private String destination;
+	
+	@ApiParam(value = "The value type to match")
+	private RelationshipValueType valueType;
+
+	@ApiParam(value = "The value comparison operator")
+	private SearchResourceRequest.Operator operator;
+
+	@ApiParam(value = "The value to match (in literal form)")
+	private String value;
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedRelationshipRestSearch.java:57]: Multi-value support?

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java [hunk 2]
```diff
@@ -487,6 +495,169 @@ public SnomedEclRefinementEvaluator(EclExpression focusConcepts) {
...
+	private Promise<Collection<Property>> evalStatementsWithValue(
+		final BranchContext context, 
+		final Set<String> focusConceptIds,
+		final Collection<String> typeIds, 
+		final RelationshipValue value, 
+		final SearchResourceRequest.Operator operator) {
+
+		// TODO: does this request need to support filtering by group?
+		Promise<Collection<Property>> relationships = SnomedRequests.prepareSearchRelationship()
+				.all()
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java:567]: To reduce memory-pressure and improve speed, use field selection here to reduce the number of fields loaded from the concrete value relationship. Same as loading fields during standard relationship search.

**Comment by apeteri** [comment]
Ready for review; the commits added today are the results of a self-review I have personally conducted. 😄

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedRelationshipSearchRequestBuilder.java [hunk 3]
```diff
@@ -86,9 +88,36 @@ public SnomedRelationshipSearchRequestBuilder filterByModifier(Iterable<String>
...
+	
+	public SnomedRelationshipSearchRequestBuilder filterByValueType(Iterable<RelationshipValueType> valueTypes) {
+		return addOption(OptionKey.VALUE_TYPE, valueTypes);
+	}
+
+	public SnomedRelationshipSearchRequestBuilder filterByValue(RelationshipValue value) {
+		return filterByValue(SearchResourceRequest.Operator.EQUALS, value);
+	}
+	
+	public SnomedRelationshipSearchRequestBuilder filterByValue(Iterable<RelationshipValue> values) {
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedRelationshipSearchRequestBuilder.java]: It is still a bit unsettled, but we have started using pluralized method names (`filterByValues`) for multi-valued filters to keep them separated from the singular versions.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedRelationshipUpdateRequest.java [hunk 4]
```diff
@@ -15,80 +15,98 @@
...
+	// Not @NotEmpty, update of type ID is optional
+	private String typeId;
+
+	// Not @NotEmpty, update of destination ID is optional
+	private String destinationId;
+
+	// Not @NotNull, update of relationship value is optional
+	private RelationshipValue value;
+
+	@Min(0)	@Max(Integer.MAX_VALUE)
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedRelationshipUpdateRequest.java]: A bit of inconsistency between the create and update requests. Create allows only `Byte.MAX_VALUE` while update allows `Integer.MAX_VALUE` `groups` and `unionGroups`.

**Review by cmark — CHANGES_REQUESTED** [review]

File: snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedRelationshipRestSearch.java (hunk 1 above)
Comment (apeteri) [reply snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedRelationshipRestSearch.java:57]: I'd keep it single valued for now. Multi-values are not great on the request level either.

**Comment by apeteri** [comment]
I've made the requested changes.

**Review by cmark — APPROVED** [review]
LGTM!

**Pull request merged**